internal/events/block_confirmations.go

```diff
@@ -0,0 +1,520 @@
+				if notification.nType == bcmStopStream {
+					// Handle stream notifications immediately
+					bcm.streamStopped(notification)
+				} else {
+					// Defer until after we've got new logs
+					notifications = append(notifications, notification)
+				}
+			}
+		}
+		pollTimer = time.NewTimer(bcm.pollingInterval)
+
+		// Setup a filter if we're missing one
+		if bcm.filterStale {
+			if err := bcm.createBlockFilter(); err != nil {
+				bcm.log.Errorf("Failed to create block filter: %s", err)
+				continue
+			}
+
+			if err := bcm.walkChain(); err != nil {
+				bcm.log.Errorf("Failed to create walk chain after restoring filter: %s", err)
+				continue
+			}
+		}
+
+		// Do the poll
+		blockHashes, err := bcm.pollBlockFilter()
+		if err != nil {
+			bcm.log.Errorf("Failed to retrieve blocks from filter: %s", err)
+			continue
+		}
+
+		// Process each new block
+		bcm.processBlockHashes(blockHashes)
+
+		// Process any new notifications - we do this at the end, so it can benefit
+		// from knowing the latest highestBlockSeen
+		if err := bcm.processNotifications(notifications); err != nil {
+			bcm.log.Errorf("Failed processing notifications: %s", err)
+			continue
+		}
+
+		// Clear the notifications array now we've processed them (we keep the slice memory)
+		notifications = notifications[:0]
+
+	}
+
+}
+
+func (bcm *blockConfirmationManager) keyForEvent(event *eventData) string {
+	return fmt.Sprintf("TX:%s|BLOCK:%s/%s|INDEX:%s|LOG:%s", event.TransactionHash, event.BlockNumber, event.BlockHash, event.TransactionIndex, event.LogIndex)
+}
+
+func (bcm *blockConfirmationManager) processNotifications(notifications []*bcmNotification) error {
+
+	for _, n := range notifications {
+		switch n.nType {
+		case bcmNewLog:
+			pending := bcm.addEvent(n.event, n.eventStream)
+			if err := bcm.walkChainForEvent(pending); err != nil {
+				return err
+			}
+		case bcmRemovedLog:
+			bcm.removeEvent(n.event)
+		default:
+			// Note that streamStopped is handled in the polling loop directly
+			bcm.log.Warnf("Unexpected notification type: %d", n.nType)
+		}
+	}
+
+	return nil
+}
+
+// streamStopped removes all pending work for a given stream, and notifies once done
+func (bcm *blockConfirmationManager) streamStopped(notification *bcmNotification) {
+	for eventKey, pending := range bcm.pending {
+		if pending.eventStream == notification.eventStream {
+			delete(bcm.pending, eventKey)
+		}
+	}
+	close(notification.complete)
+}
+
+// addEvent is called by the goroutine on receipt of a new event notification
+func (bcm *blockConfirmationManager) addEvent(event *eventData, eventStream *eventStream) *pendingEvent {
+
+	// Add the event
+	eventKey := bcm.keyForEvent(event)
+	pending := &pendingEvent{
+		key:           bcm.keyForEvent(event),
+		added:         time.Now(),
+		confirmations: make([]*blockInfo, 0, bcm.requiredConfirmations),
+		event:         event,
+		eventStream:   eventStream,
+	}
+	bcm.pending[eventKey] = pending
```

Comment:
Not requesting as part of this PR given I'm providing this comment at the last minute, it would be good to have a follow on that wraps updates to `bcm.pending` with a RWMutex - I'm assuming there's a high likelihood of concurrent updates.

Reply:
Agree a good candidate to look at in the future, and deferrable.